cpp/ql/src/Critical/MissingCheckScanf.ql

```diff
@@ -0,0 +1,123 @@
+/**
+ * @name Missing return-value check for a 'scanf'-like function
+ * @description Without checking that a call to 'scanf' actually wrote to an
+ *              output variable, reading from it can lead to unexpected behavior.
+ * @kind problem
+ * @problem.severity recommendation
+ * @security-severity 4.5
+ * @precision medium
+ * @id cpp/missing-check-scanf
+ * @tags security
+ */
+
+import cpp
+import semmle.code.cpp.commons.Scanf
+import semmle.code.cpp.controlflow.Guards
+import semmle.code.cpp.dataflow.DataFlow
+import semmle.code.cpp.ir.IR
+import semmle.code.cpp.ir.ValueNumbering
+
+/** An expression appearing as an output argument to a `scanf`-like call */
+class ScanfOutput extends Expr {
+  ScanfFunctionCall call;
+  int varargIndex;
+  Instruction instr;
+  ValueNumber valNum;
+
+  ScanfOutput() {
+    this = call.getArgument(call.getTarget().getNumberOfParameters() + varargIndex) and
+    varargIndex >= 0 and
```

Comment:
Yeah, I was seriously considering moving part of this ScanfOutput class into a the Scanf.qll library. `getAnOutputArgument()` would be a useful function, as well as, potentially, `getMinimumGuardConstant()`, due to having special cases that you don't want duplicated elsewhere (assuming other queries are interested in this information).

Reply:
I agree it would be nice to have `ScanfFunctionCall.getAnOutputArgument()` (or `ScanfFunctionCall.getOutputArgument(int index)`, or both).  I'm not entirely clear what `getMinimumGuardConstant()` would be exactly, is this the number of vararg arguments that should have been written (and thus also the return value) if the `scanf` is entirely successful?

Reply:
Yes, it's the smallest return value that indicates success in writing the given argument.

Reply:
Btw, is there a convention that "parameter" refers to the variable in the parameter list from the perspective of the body of a function, while "argument" refers to an expression passed at the call site of a function?

Reply:
Yep.